core/internal/messagelog/messagelog.go

```diff
@@ -37,38 +37,44 @@ import (
 // they appear in the log. Closing the channel passed to this function
 // indicates the returned channel should be closed. Nil channel may be
 // passed if there's no need to close the returned channel.
+//
+// Messages returns all messages currently in the log.
+//
+// Reset replaces messages stored in the log with the supplied ones.
 type MessageLog interface {
 	Append(msg messages.Message)
 	Stream(done <-chan struct{}) <-chan messages.Message
+	Messages() []messages.Message
+	Reset(msgs []messages.Message)
 }
 
 type messageLog struct {
-	lock sync.RWMutex
+	sync.RWMutex
+
+	// Value incremented on each reset
+	epoch uint64
```

Comment:
This field is used to detect `Reset()` calls during `supplyMessages()`. So naming this more directly like "reset" might be suitable? I think that the word "epoch" can be used to represent the similar concept as "view", so it might be confusing for someone familiar with other consensus protocols.

Reply:
I decided to replace it with a channel.